compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaLoopSubsumption.cpp

```diff
@@ -0,0 +1,402 @@
+/// 'other' operation.
+Operation *getParentInSameBlock(Operation *op, Operation *other) {
+  if (!op || !other) return nullptr;
+  if (op->getBlock() == other->getBlock()) return op;
+  Operation *parentOp = op->getParentOp();
+  return getParentInSameBlock(parentOp, other);
+}
+
+/// Utility affine expression visitor to retrieve the stride from the
+/// expression.
+struct RetrieveStrideSize : public AffineExprVisitor<RetrieveStrideSize> {
+  std::optional<int64_t> stride;
+  RetrieveStrideSize() {}
+  void visitMulExpr(AffineBinaryOpExpr expr) {
+    if (auto rhsSize = dyn_cast<AffineConstantExpr>(expr.getRHS())) {
+      stride = rhsSize.getValue();
+    } else if (auto lhsSize = dyn_cast<AffineConstantExpr>(expr.getLHS())) {
+      stride = lhsSize.getValue();
+    }
+  }
+};
+
+/// Utility to clean up the DMA users after loop subsumption + hoisting. This
+/// will hoist `amdaie.npu.dma_cpy_nd`'s users like `npu.dma_wait` as well.
+LogicalResult moveUsersToHoistedDMAScope(Operation *parentOp) {
+  IRRewriter rewriter(parentOp->getContext());
+  // Move `amdaie.npu.dma_wait` operation after the parent op in the same block
+  // as the input `amdaie.npu.dma_cpy_nd` operation. This parent op will
+  // typically be a loop out of which the DMA operation has been hoisted. Moving
+  // the wait operation after this loop is important to avoid a deadlock with
+  // whatever operations are still remaining inside the loop's scope.
+  WalkResult res = parentOp->walk([&](AMDAIE::NpuDmaWaitOp npuDmaWaitOp) {
+    Operation *dmaOp = npuDmaWaitOp.getDma().getDefiningOp();
+    Operation *parentInSameBlock = getParentInSameBlock(npuDmaWaitOp, dmaOp);
+    if (!parentInSameBlock) return WalkResult::interrupt();
+    rewriter.moveOpAfter(npuDmaWaitOp, parentInSameBlock);
+    return WalkResult::advance();
+  });
+  if (res.wasInterrupted()) return failure();
+  return success();
+}
+
+class SubsumeLoopIntoDMA
+    : public OpInterfaceRewritePattern<AMDAIE::DoublyStridedOpInterface> {
+  using OpInterfaceRewritePattern::OpInterfaceRewritePattern;
+
+  /// Utility to add a loop iteration to an offsets/sizes/strides access
+  /// pattern.
+  LogicalResult addIterationToAccessPattern(
+      RewriterBase &rewriter, int64_t lowerBound, int64_t upperBound,
+      int64_t step, const DenseSet<Value> &applyValues,
+      SmallVector<OpFoldResult> &newOffsets,
+      SmallVector<OpFoldResult> &newSizes,
+      SmallVector<OpFoldResult> &newStrides) const {
+    SmallVector<OpFoldResult> insertOffsets;
+    SmallVector<OpFoldResult> insertSizes;
+    SmallVector<OpFoldResult> insertStrides;
+    for (auto &&[i, offset] : llvm::enumerate(newOffsets)) {
+      Value offsetValue = getValueOrCreateConstantIndexOp(
+          rewriter, rewriter.getUnknownLoc(), offset);
+      if (applyValues.contains(offsetValue)) {
+        auto applyOp =
+            dyn_cast<affine::AffineApplyOp>(offsetValue.getDefiningOp());
+        if (!applyOp) return failure();
```

Comment:
I was thinking about the case where the offset would be an induction variable instead of an `affine.apply` operating on an induction variable, but then I didn't handle that case as we don't see it right now and as the PR was becoming rather large already, I wanted to handle that in a follow-up. But I added it to this PR itself now.